<HTML>
<BODY>


<H4> getPairAligns - Get all the contigs pairwise alignments </H4>
<UL>
<LI> getPairAlignsC2.m - symmetry requirement
<LI> getPairAlignsCe.m - experimental data
<LI> getPairAligns.m - non-contigs' version (OLD)
<LI> getPairAlignsC.m - no symmetry requirement version (OLD)
<BR><BR>
<LI> init.m - Starts a work session for the petide spectra alignment project
<LI> getScoreDistContigs.m - Generates a dataset for estimating the relevance of a score given (shift, aligned peptide's mass)
<LI> contigRanges.m - merge the locations of the peptides in a contig (OLD)
<LI> contigRanges2.m (OLD)
</UL>

<H4> Data </H4>
<UL>
<LI>AAdata.dat - Mono-isotopic amino acid masses
<LI>AAmassesAvg.mat - Average amino acid masses
<LI>AAdata_oldMasses.dat - Old table of mono-isotopic amino acid masses
<BR><BR>
<LI>msProfile.mat - Dancik'99 ion probabilities with integer offsets (added B2/b2-H20)
<LI>msProfileReal.mat - Dancik'99 ion probabilities with real offsets
<LI>msProfileISB.mat - ISB ion probabilities with real offsets + tolerance of 0.5 Da
<LI>msProfile_old.mat - Dancik'99 ion probabilities with integer offsets
<BR><BR>
<LI>data_spectrin.mat - spectrin fragment positions + protein sequence
<LI>proteinData.dat - P11217, P16278, Q9UBC5 amino acid sequences
</UL>

</BODY>
</HTML>
